### ISSUE TITLE
Enable verifying proofs in parallel, in similar or in different modules, using a single thread and solver pool

### ISSUE THREAD

**Issue by keyboardDrummer** [dafny-lang/dafny#1647]
Dafny currently has the options `vcsCores` and `vcsLoad` that will let it verify proofs that occur inside the same module, in parallel, although it will still verify proofs in separate modules sequentially.

When the `vcsCores` or `vcsLoad` option is used, Dafny should also verify proofs in different modules in parallel, and it should not require additional thread or checker pools for this.

### Out of scope

Parsing files and resolving or compiling modules in parallel is out of scope for this story

### Implementation suggestions

- Remove references to `CommandLineOptions.Clo` in Boogie (392 in total). This requires chaining an options object through the Boogie call graph. This requires a lot of simple code changes and is, I expect, the bulk of the work.
- Remove references to `Console.Out` in Boogie (57 in total). This is simple once an options object is already chained through the Boogie call graph.
- Change Boogie's `InferAndVerify` so it returns an `IReadOnlyList<ProofExecution>`. A `ProofExecution` represents a verification condition. It contains all the information needed to identify a verification condition and a `Task` that can used to track the state of this verification. `ExecutionEngine.VerifyImplementation` and `ConditionGeneriation.VerifyImplementation` must also be changed to return a `IReadOnlyList<ProofExecution>`
- Change Boogie's `InferAndVerify` so it takes an additional `CheckerPool` argument. Change Dafny to create a single CheckerPool which it passes in all invocations of `InferAndVerify`.
- In Dafny, thread Boogie's Task objects up the call chain to `DafnyDriver.Boogie`, and there wait for all of them to complete. 
- A reporting class that can handle reporting verification progress for different Boogie modules concurrently, must be added to Dafny.
- If needed, configure a ThreadPoolScheduler in Dafny that creates threads with increased memory to prevent stack overflows.

**Comment by keyboardDrummer** [comment]
Closed with https://github.com/dafny-lang/dafny/pull/1870
